diff -rupN systemc-2.3.0/src/sysc/kernel/sc_simcontext.h systemc-2.3.0.modified/src/sysc/kernel/sc_simcontext.h
--- systemc-2.3.0/src/sysc/kernel/sc_simcontext.h	2011-09-05 15:20:22.000000000 -0600
+++ systemc-2.3.0.modified/src/sysc/kernel/sc_simcontext.h	2014-06-11 13:54:17.166545025 -0600
@@ -348,7 +348,9 @@ private:
     sc_runnable*                m_runnable;
     sc_process_list*            m_collectable;
 
+public:
     sc_time_params*             m_time_params;
+private:
     sc_time                     m_curr_time;
     mutable sc_time             m_max_time;
