--- original/qsim-0.2.1/mgzd.h	2013-08-08 16:10:15.000000000 -0600
+++ patched/qsim-0.2.1/mgzd.h	2013-09-24 14:24:42.148558727 -0600
@@ -40,16 +40,37 @@
     lib.file = tmp_filename_ptr;
     free((void *)tmp_filename_ptr);
 
-    std::ostringstream cp_command;
+    /* std::ostringstream cp_command; */
 
-    cp_command << "cp " << libfile << ' ' << lib.file;
-    int r;
-    if ((r = system(cp_command.str().c_str())) != 0) {
-      std::cerr << "system(\"" << cp_command.str() 
-                << "\") returned " << r <<".\nrm /";
+    /* cp_command << "cp " << libfile << ' ' << lib.file; */
+    /* int r; */
+    /* if ((r = system(cp_command.str().c_str())) != 0) { */
+    /*   std::cerr << "system(\"" << cp_command.str()  */
+    /*             << "\") returned " << r <<".\nrm /"; */
+    /*   exit(1); */
+    /* } */
+
+    FILE* original_lib = fopen(libfile, "rb");
+    if(original_lib == NULL || ferror(original_lib)) {
+      perror("Unable to open qsim library.");
       exit(1);
     }
 
+    FILE* copy_lib     = fopen(lib.file.c_str(), "wb");
+    if(copy_lib == NULL || ferror(copy_lib)) {
+      perror("Unable to open the target to copy qsim library to");
+      exit(1);
+    }
+
+    char* copy_buffer = (char*) malloc(sizeof(char) * 2048);
+    while(! feof(original_lib) ) {
+      size_t copied_size = fread(copy_buffer, sizeof(char), 2048, original_lib);
+      fwrite(copy_buffer, sizeof(char), copied_size, copy_lib);
+    }
+
+    fclose(original_lib);
+    fclose(copy_lib);
+
     lib.handle = dlopen(lib.file.c_str(), RTLD_NOW|RTLD_LOCAL);
     if (lib.handle == NULL) {
       std::cerr << "dlopen(\"" << lib.file.c_str() << "\") failed:  " 
